### PR TITLE
JDK-8301223: Replace NULL with nullptr in share/gc/g1/

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi, this PR changes all occurrences of NULL to nullptr for the subdirectory share/gc/g1/. Unfortunately the script that does the change isn't perfect, and so we
need to comb through these manually to make sure nothing has gone wrong. I also review these changes but things slip past my eyes sometimes.

Here are some typical things to look out for:

1. No changes but copyright header changed (probably because I reverted some changes but forgot the copyright).
2. Macros having their NULL changed to nullptr, these are added to the script when I find them. They should be NULL.
3. nullptr in comments and logs. We try to use lower case "null" in these cases as it reads better. An exception is made when code expressions are in a comment.

An example of this:

```c++
// This function returns null
void* ret_null();
// This function returns true if *x == nullptr
bool is_nullptr(void** x);
```

Note how `nullptr` participates in a code expression here, we really are talking about the specific value `nullptr`.

Thanks!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301223](https://bugs.openjdk.org/browse/JDK-8301223): Replace NULL with nullptr in share/gc/g1/

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [9bb7edbe](https://git.openjdk.org/jdk/pull/12248/files/9bb7edbecfaa0ea71d84737ab25dbb6b66d73f49)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to [9bb7edbe](https://git.openjdk.org/jdk/pull/12248/files/9bb7edbecfaa0ea71d84737ab25dbb6b66d73f49)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12248/head:pull/12248` \
`$ git checkout pull/12248`

Update a local copy of the PR: \
`$ git checkout pull/12248` \
`$ git pull https://git.openjdk.org/jdk.git pull/12248/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12248`

View PR using the GUI difftool: \
`$ git pr show -t 12248`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12248.diff">https://git.openjdk.org/jdk/pull/12248.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12248#issuecomment-1512803096)

### CODE REVIEW

File: src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp [hunk 1]
```diff
@@ -73,12 +73,12 @@ const TypeFunc *G1BarrierSetC2::write_ref_field_post_entry_Type() {
  * marking are kept alive, all reference updates need to any previous
  * reference stored before writing.
  *
- * If the previous value is NULL there is no need to save the old value.
- * References that are NULL are filtered during runtime by the barrier
+ * If the previous value is nullptr there is no need to save the old value.
+ * References that are nullptr are filtered during runtime by the barrier
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp:77]:
```suggestion
 * If the previous value is null there is no need to save the old value.
 * References that are null are filtered during runtime by the barrier
```

File: src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp [hunk 2]
```diff
@@ -73,12 +73,12 @@ const TypeFunc *G1BarrierSetC2::write_ref_field_post_entry_Type() {
...
  * code to avoid unnecessary queuing.
  *
  * However in the case of newly allocated objects it might be possible to
- * prove that the reference about to be overwritten is NULL during compile
+ * prove that the reference about to be overwritten is nullptr during compile
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp:81]:
```suggestion
 * prove that the reference about to be overwritten is null during compile
```

File: src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp [hunk 3]
```diff
@@ -248,9 +248,9 @@ void G1BarrierSetC2::pre_barrier(GraphKit* kit,
       pre_val = __ load(__ ctrl(), adr, val_type, bt, alias_idx, false, MemNode::unordered, LoadNode::Pinned);
     }
 
-    // if (pre_val != NULL)
+    // if (pre_val != null)
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp:251]:
```suggestion
    // if (pre_val != nullptr)
```
This is converted wrongly in a comment

File: src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp [hunk 4]
```diff
@@ -248,9 +248,9 @@ void G1BarrierSetC2::pre_barrier(GraphKit* kit,
...
     __ if_then(pre_val, BoolTest::ne, kit->null()); {
-      Node* buffer  = __ load(__ ctrl(), buffer_adr, TypeRawPtr::NOTNULL, T_ADDRESS, Compile::AliasIdxRaw);
+      Node* buffer  = __ load(__ ctrl(), buffer_adr, TypeRawPtr::NOTnullptr, T_ADDRESS, Compile::AliasIdxRaw);
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp:253]:
```suggestion
      Node* buffer  = __ load(__ ctrl(), buffer_adr, TypeRawPtr::NOTNULL, T_ADDRESS, Compile::AliasIdxRaw);
```

File: src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp [hunk 5]
```diff
@@ -270,7 +270,7 @@ void G1BarrierSetC2::pre_barrier(GraphKit* kit,
         const TypeFunc *tf = write_ref_field_pre_entry_Type();
         __ make_leaf_call(tf, CAST_FROM_FN_PTR(address, G1BarrierSetRuntime::write_ref_field_pre_entry), "write_ref_field_pre_entry", pre_val, tls);
       } __ end_if();  // (!index)
-    } __ end_if();  // (pre_val != NULL)
+    } __ end_if();  // (pre_val != null)
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp:273]:
```suggestion
    } __ end_if();  // (pre_val != nullptr)
```

File: src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp [hunk 6]
```diff
@@ -42,8 +42,8 @@
 
 const TypeFunc *G1BarrierSetC2::write_ref_field_pre_entry_Type() {
   const Type **fields = TypeTuple::fields(2);
-  fields[TypeFunc::Parms+0] = TypeInstPtr::NOTNULL; // original field value
-  fields[TypeFunc::Parms+1] = TypeRawPtr::NOTNULL; // thread
+  fields[TypeFunc::Parms+0] = TypeInstPtr::NOTnullptr; // original field value
+  fields[TypeFunc::Parms+1] = TypeRawPtr::NOTnullptr; // thread
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp:46]:
```suggestion
  fields[TypeFunc::Parms+0] = TypeInstPtr::NOTNULL; // original field value
  fields[TypeFunc::Parms+1] = TypeRawPtr::NOTNULL; // thread
```

File: src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp [hunk 7]
```diff
@@ -55,8 +55,8 @@ const TypeFunc *G1BarrierSetC2::write_ref_field_pre_entry_Type() {
 
 const TypeFunc *G1BarrierSetC2::write_ref_field_post_entry_Type() {
   const Type **fields = TypeTuple::fields(2);
-  fields[TypeFunc::Parms+0] = TypeRawPtr::NOTNULL;  // Card addr
-  fields[TypeFunc::Parms+1] = TypeRawPtr::NOTNULL;  // thread
+  fields[TypeFunc::Parms+0] = TypeRawPtr::NOTnullptr;  // Card addr
+  fields[TypeFunc::Parms+1] = TypeRawPtr::NOTnullptr;  // thread
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp:59]:
```suggestion
  fields[TypeFunc::Parms+0] = TypeRawPtr::NOTNULL;  // Card addr
  fields[TypeFunc::Parms+1] = TypeRawPtr::NOTNULL;  // thread
```

File: src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp [hunk 8]
```diff
@@ -288,7 +288,7 @@ void G1BarrierSetC2::pre_barrier(GraphKit* kit,
  *
  * To reduce the number of updates to the remembered set the post-barrier
  * filters updates to fields in objects located in the Young Generation,
- * the same region as the reference, when the NULL is being written or
+ * the same region as the reference, when the nullptr is being written or
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp:291]:
```suggestion
 * the same region as the reference, when null is being written or
```

File: src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp [hunk 9]
```diff
@@ -377,13 +377,13 @@ void G1BarrierSetC2::post_barrier(GraphKit* kit,
                                   Node* val,
                                   BasicType bt,
                                   bool use_precise) const {
-  // If we are writing a NULL then we need no post barrier
+  // If we are writing a null then we need no post barrier
 
-  if (val != NULL && val->is_Con() && val->bottom_type() == TypePtr::NULL_PTR) {
-    // Must be NULL
+  if (val != nullptr && val->is_Con() && val->bottom_type() == TypePtr::nullptr_PTR) {
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp:382]:
```suggestion
  if (val != nullptr && val->is_Con() && val->bottom_type() == TypePtr::NULL_PTR) {
```

File: src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp [hunk 10]
```diff
@@ -377,13 +377,13 @@ void G1BarrierSetC2::post_barrier(GraphKit* kit,
...
+    // Must be null
     const Type* t = val->bottom_type();
-    assert(t == Type::TOP || t == TypePtr::NULL_PTR, "must be NULL");
-    // No post barrier if writing NULLx
+    assert(t == Type::TOP || t == TypePtr::nullptr_PTR, "must be null");
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp:385]:
```suggestion
    assert(t == Type::TOP || t == TypePtr::NULL_PTR, "must be null");
```

File: src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp [hunk 11]
```diff
@@ -377,13 +377,13 @@ void G1BarrierSetC2::post_barrier(GraphKit* kit,
...
+    // No post barrier if writing nullx
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp:386]:
```suggestion
    // No post barrier if writing null.
```

File: src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp [hunk 12]
```diff
@@ -434,7 +434,7 @@ void G1BarrierSetC2::post_barrier(GraphKit* kit,
   // Use ctrl to avoid hoisting these values past a safepoint, which could
   // potentially reset these fields in the JavaThread.
   Node* index  = __ load(__ ctrl(), index_adr, TypeX_X, TypeX_X->basic_type(), Compile::AliasIdxRaw);
-  Node* buffer = __ load(__ ctrl(), buffer_adr, TypeRawPtr::NOTNULL, T_ADDRESS, Compile::AliasIdxRaw);
+  Node* buffer = __ load(__ ctrl(), buffer_adr, TypeRawPtr::NOTnullptr, T_ADDRESS, Compile::AliasIdxRaw);
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp:437]:
```suggestion
  Node* buffer = __ load(__ ctrl(), buffer_adr, TypeRawPtr::NOTNULL, T_ADDRESS, Compile::AliasIdxRaw);
```

File: src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp [hunk 13]
```diff
@@ -459,7 +459,7 @@ void G1BarrierSetC2::post_barrier(GraphKit* kit,
     // if (xor_res == 0) same region so skip
     __ if_then(xor_res, BoolTest::ne, zeroX, likely); {
 
-      // No barrier if we are storing a NULL
+      // No barrier if we are storing a null
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp:462]:
```suggestion
      // No barrier if we are storing a null.
```

File: src/hotspot/share/gc/g1/g1CollectedHeap.hpp [hunk 14]
```diff
@@ -339,7 +339,7 @@ class G1CollectedHeap : public CollectedHeap {
 #define assert_at_safepoint_on_vm_thread()                                    \
   do {                                                                        \
     assert_at_safepoint();                                                    \
-    assert(Thread::current_or_null() != NULL, "no current thread");           \
+    assert(Thread::current_or_null() != nullptr, "no current thread");           \
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1CollectedHeap.hpp:342]:
```suggestion
    assert(Thread::current_or_null() != nullptr, "no current thread");        \
```

File: src/hotspot/share/gc/g1/g1CollectedHeap.hpp [hunk 15]
```diff
@@ -506,7 +506,7 @@ class G1CollectedHeap : public CollectedHeap {
   // Attempting to expand the heap sufficiently
   // to support an allocation of the given "word_size".  If
   // successful, perform the allocation and return the address of the
-  // allocated block, or else "NULL".
+  // allocated block, or else "null".
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1CollectedHeap.hpp:509]:
```suggestion
  // allocated block, or else null.
```

File: src/hotspot/share/gc/g1/g1ConcurrentMark.cpp [hunk 16]
```diff
@@ -1918,7 +1918,7 @@ HeapRegion* G1ConcurrentMark::claim_region(uint worker_id) {
       } else {
         assert(limit == bottom,
                "the region limit should be at bottom");
-        // we return NULL and the caller should try calling
+        // we return null and the caller should try calling
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1ConcurrentMark.cpp:1921]:
```suggestion
        // We return null and the caller should try calling
```

File: src/hotspot/share/gc/g1/g1FreeIdSet.cpp [hunk 17]
```diff
@@ -33,7 +33,7 @@
 
 G1FreeIdSet::G1FreeIdSet(uint start, uint size) :
   _sem(size),          // counting semaphore for available ids
-  _next(NULL),         // array of "next" indices
+  _next(nullptr),         // array of "next" indices
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1FreeIdSet.cpp:36]:
```suggestion
  _next(nullptr),      // array of "next" indices
```

File: src/hotspot/share/gc/g1/g1RemSet.cpp [hunk 18]
```diff
@@ -1492,14 +1492,14 @@ bool G1RemSet::clean_card_before_refine(CardValue** const card_ptr_addr) {
...
-  if (r == NULL) {
+  if (r == nullptr) {
     return false;
   }
 
   check_card_ptr(card_ptr, _ct);
 
   // If the card is no longer dirty, nothing to do.
-  // We cannot load the card value before the "r == NULL" check, because G1
+  // We cannot load the card value before the "r == null" check, because G1
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1RemSet.cpp:1502]:
```suggestion
  // We cannot load the card value before the "r == nullptr" check above, because G1
```

**Review by tschatzl — CHANGES_REQUESTED** [review]

**Comment by jdksjolen** [comment]
Hm, got some funny errors:

```

=== Output from failing command(s) repeated here ===
* For target hotspot_variant-server_libjvm_objs_filemap.o:
/home/runner/work/jdk/jdk/src/hotspot/share/cds/filemap.cpp: In member function 'void FileMapInfo::seek_to_position(size_t)':
/home/runner/work/jdk/jdk/src/hotspot/share/cds/filemap.cpp:1429:50: error: format '%ld' expects argument of type 'long int', but argument 2 has type 'size_t' {aka 'unsigned int'} [-Werror=format=]
 1429 |     log_error(cds)("Unable to seek to position %ld", pos);
      |                                                ~~^   ~~~
      |                                                  |   |
      |                                                  |   size_t {aka unsigned int}
      |                                                  long int
      |                                                %d
cc1plus: all warnings being treated as errors
```

I'll have to look into the merge and see if I messed that up. Worst case, I'll re-apply the nullptr conversion to origin/master.

**Comment by kimbarrett** [comment]
> Hm, got some funny errors:
> 
> ```
> 
> === Output from failing command(s) repeated here ===
> * For target hotspot_variant-server_libjvm_objs_filemap.o:
> /home/runner/work/jdk/jdk/src/hotspot/share/cds/filemap.cpp: In member function 'void FileMapInfo::seek_to_position(size_t)':
> /home/runner/work/jdk/jdk/src/hotspot/share/cds/filemap.cpp:1429:50: error: format '%ld' expects argument of type 'long int', but argument 2 has type 'size_t' {aka 'unsigned int'} [-Werror=format=]
>  1429 |     log_error(cds)("Unable to seek to position %ld", pos);
>       |                                                ~~^   ~~~
>       |                                                  |   |
>       |                                                  |   size_t {aka unsigned int}
>       |                                                  long int
>       |                                                %d
> cc1plus: all warnings being treated as errors
> ```
> 
> I'll have to look into the merge and see if I messed that up. Worst case, I'll re-apply the nullptr conversion to origin/master.

That seems to be  https://bugs.openjdk.org/browse/JDK-8306289, which has been fixed.

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Review by tschatzl — APPROVED** [review]
There is one more `NULL` in
```

HeapWord*
G1CollectedHeap::humongous_obj_allocate_initialize_regions(HeapRegion* first_hr,
                                                           uint num_regions,
                                                           size_t word_size) {
  assert(first_hr != NULL, "pre-condition");
```
line 281. I do not seem to be able to anchor this suggestion in the github UI.

Looks good otherwise.

**Comment by jdksjolen** [comment]
Fixed the remaining nullptr conversion that tschatzl pointed out and Mach5 passes tier1.

/integrate

**Pull request closed without merging**